### PR TITLE
Write Registrar worker configuration file

### PR DESCRIPTION
**Pull request by kdmccormick**
@edx/masters-neem 
https://openedx.atlassian.net/servicedesk/customer/portal/3/DEVOPS-8691
https://openedx.atlassian.net/browse/EDUCATOR-4301

Configuration Pull Request
---

Make sure that the following steps are done before merging:

  - [ ] A DevOps team member has approved the PR if it is code shared across multiple services and you don't own all of the services.
  - [ ] Are you adding any new default values that need to be overridden when this change goes live? If so:
    - [ ] Update the appropriate internal repo (be sure to update for all our environments)
    - [ ] If you are updating a secure value rather than an internal one, file a DEVOPS ticket with details.
    - [x] Add an entry to the CHANGELOG.
  - [ ] If you are making a complicated change, have you performed the proper testing specified on the [Ops Ansible Testing Checklist](https://openedx.atlassian.net/wiki/display/EdxOps/Ops+Ansible+Testing+Checklist)?  Adding a new variable does not require the full list (although testing on a sandbox is a great idea to ensure it links with your downstream code changes).
  - [ ] Think about how this change will affect Open edX operators.  Have you updated the wiki page for the next Open edX release?

### CODE REVIEW

File: playbooks/roles/registrar/templates/edx/app/supervisor/conf.d.available/registrar-workers.conf.j2 [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{% for w in registrar_workers %}
```
Comment (fredsmith) [inline playbooks/roles/registrar/templates/edx/app/supervisor/conf.d.available/registrar-workers.conf.j2:1]: unless you've written something like this (https://github.com/edx/configuration/blob/b195d878a64afb6afb85e6ab231a83ead3406443/playbooks/roles/edxapp/defaults/main.yml#L607) into the defaults, this isn't going to work.

you should probably scan below to see what parts of 'w' are relevant to your workers and add the defaults

File: playbooks/roles/supervisor/files/pre_supervisor_checks.py [hunk 2]
```diff
@@ -226,7 +235,8 @@ def edp_for_instance(instance_id):
                     "ecommerce": {'python': args.ecommerce_python, 'env_file': args.ecommerce_env, 'code_dir': args.ecommerce_code_dir},
                     "credentials": {'python': args.credentials_python, 'env_file': args.credentials_env, 'code_dir': args.credentials_code_dir},
                     "discovery": {'python': args.discovery_python, 'env_file': args.discovery_env, 'code_dir': args.discovery_code_dir},
-                    "journals": {'python': args.journals_python, 'env_file': args.journals_env, 'code_dir': args.journals_code_dir},                    
+                    "journals": {'python': args.journals_python, 'env_file': args.journals_env, 'code_dir': args.journals_code_dir},
```
Comment (fredsmith) [inline playbooks/roles/supervisor/files/pre_supervisor_checks.py]: did spacing here change?

File: playbooks/roles/registrar/templates/edx/app/supervisor/conf.d.available/registrar-workers.conf.j2 [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+user={{ common_web_user }}
+directory={{ registrar_code_dir }}
+stdout_logfile={{ supervisor_log_dir }}/%(program_name)s-stdout.log
+stderr_logfile={{ supervisor_log_dir }}/%(program_name)s-stderr.log
+
+command={{ registrar_app_dir }}/worker.sh {{ w.service_variant }} --settings={{ worker_django_settings_module }} celery worker --loglevel=info --queues=edx.{{ w.service_variant }}.core.{{ w.queue }} --hostname=edx.{{ w.service_variant }}.core.{{ w.queue }}.%%h --concurrency={{ w.concurrency }} {{ '--maxtasksperchild ' + w.max_tasks_per_child|string if w.max_tasks_per_child is defined else '' }} {{ '--without-heartbeat' if not REGISTRAR_CELERY_HEARTBEAT_ENABLED|bool else '' }}
+killasgroup=true
+stopwaitsecs={{ w.stopwaitsecs | default(REGISTRAR_WORKER_DEFAULT_STOPWAITSECS) }}
```
Comment (fredsmith) [inline playbooks/roles/registrar/templates/edx/app/supervisor/conf.d.available/registrar-workers.conf.j2:12]: make sure you add the default value for `REGISTRAR_WORKER_DEFAULT_STOPWAITSECS` to the registrar role.

**Review by fredsmith — CHANGES_REQUESTED** [review]
see review notes

File: playbooks/roles/supervisor/files/pre_supervisor_checks.py (hunk 2 above)
Comment (kdmccormick) [reply playbooks/roles/supervisor/files/pre_supervisor_checks.py]: Trailing whitespace, i think my editor chopped it off automatically. I'll put it back to reduce diff noise

**Comment by kdmccormick** [comment]
Thanks @fredsmith . One question -- [this](https://github.com/edx/configuration/pull/5128/files#diff-9fbd95606ec576d3fc445df2718f94c8R4) line sets `BOTO_CONFIG="{{ registrar_app_dir }}/.boto)"`. Will that work if `.boto` doesn't exist?

**Comment by kdmccormick** [comment]
@fredsmith do you know what this build failure means?
```
TASK [nginx : Create NGINX server templates] ***********************************
changed: [127.0.0.1] => (item={u'lang': u'en', u'img': u'https://upload.wikimedia.org/wikipedia/commons/thumb/1/11/Pendleton_Sinking_Ship.jpg/640px-Pendleton_Sinking_Ship.jpg', u'title': u'Rate limit exceeded', u'style_p': u'font-family: Georgia,Cambria,"Times New Roman",Times,serif; margin: auto; margin-bottom: 1em; font-weight: 200; line-height: 1.4em; font-size: 1.1em; max-width: 80%;', u'style_div': u'margin: auto; width: 800px; text-align: center; padding:20px 0px 0px 0px;', u'img_alt': u'', u'heading': u'Uh oh, we are having some server issues..', u'file': u'rate-limit.html', u'msg': u'If think you have encountered this message in error please let us know at <a href="mailto:technical@example.com">technical@example.com</a>', u'style_h1': u'font-family: "Helvetica Neue",Helvetica,Roboto,Arial,sans-serif; margin-bottom: .3em; font-size: 2.0em; line-height: 1.25em; text-rendering: optimizeLegibility; font-weight: bold; color: #000000;', u'style_p_h2': u'font-family: "Helvetica Neue",Helvetica,Roboto,Arial,sans-serif; margin-bottom: .3em; line-height: 1.25em; text-rendering: optimizeLegibility; font-weight: bold; font-size: 1.8em; color: #5b5e63;'})
```

**Review by fredsmith — APPROVED** [review]

**Comment by fredsmith** [comment]
@kdmccormick the build failure is: https://travis-ci.com/edx/configuration/jobs/201519083#L3744 Not the nginx template line you commented with (somewhere, I can't find it)

**Comment by fredsmith** [comment]
> Thanks @fredsmith . One question -- [this](https://github.com/edx/configuration/pull/5128/files#diff-9fbd95606ec576d3fc445df2718f94c8R4) line sets `BOTO_CONFIG="{{ registrar_app_dir }}/.boto)"`. Will that work if `.boto` doesn't exist?

It should, as long as the www-data user has write permission to the directory to create it.  I would follow the pattern used in other apps.

File: playbooks/roles/registrar/defaults/main.yml [hunk 4]
```diff
@@ -125,4 +130,17 @@ REGISTRAR_ENABLE_NEWRELIC_DISTRIBUTED_TRACING: false
 # Remote config
 REGISTRAR_HERMES_ENABLED: "{{ COMMON_HERMES_ENABLED }}"
 
+# Worker settings
+REGISTRAR_WORKERS_ENABLE_NEWRELIC_DISTRIBUTED_TRACING: false
+REGISTRAR_WORKER_DEFAULT_STOPWAITSECS: 432000
+REGISTRAR_CELERY_HEARTBEAT_ENABLED: true
+REGISTRAR_NEWRELIC_WORKERS_APPNAME: registrar-workers
```
Comment (kdmccormick) [inline playbooks/roles/registrar/defaults/main.yml:137]: @fredsmith is this a reasonable value `REGISTRAR_NEWRELIC_WORKERS_APPNAME`?

Comment (fredsmith) [reply playbooks/roles/registrar/defaults/main.yml:137]: yep

**Review by fredsmith — APPROVED** [review]

**Pull request merged**